src/dataframe/dataframe.jl

```diff
@@ -1317,3 +1317,77 @@ end
 function permutecols!(df::DataFrame, p::AbstractVector{Symbol})
     permutecols!(df, index(df)[p])
 end
+
+##############################################################################
+##
+## Expand & Complete dataframes
+##
+##############################################################################
+
+function _expandhelper(col, vect, stride, offset)
+    for i in offset:stride:length(vect)
+        push!(col, vect[i])
+    end
+end
+
+function expand(df::AbstractDataFrame, indexcols)
+    # Check to make sure the symbols in indexcols are in the df, and check for duplicate rows in the input df
+    allunique(indexcols) || throw(ArgumentError("Elements of $indexcols must be unique"))
+    sum(DataFrames.nonunique(df[:,indexcols]))>0 && @warn "duplicate rows in input; expand will only return unique combinations"
+    colind = index(df)[indexcols]
+
+    dummydf = similar(select(df, indexcols, copycols=false),0)
+
+    # Create a vect of vectors of unique values in each column
+    uniqueVals = []
+    for col in colind
+        # levels drops missing, handle the case where missing values are present
+        if sum(ismissing.(df[col])) != 0
+            tempcol = vcat(levels(df[col]), missing)
+        else
+            tempcol = levels(df[col])
+        end
+        push!(uniqueVals, tempcol)
+    end
+
+    # Get a long vector of every possible combination
+    collected = collect(flatten(product(values(uniqueVals)...)))
+    dfsize = div(length(collected),length(indexcols))
+
+    for i in 1:length(indexcols)
+        _expandhelper(dummydf[!,i], collected, length(indexcols), i)
+    end
+
+    return dummydf
+end
+
+function complete(df::AbstractDataFrame, indexcols::Array{Symbol,1}; fill=missing::Any, replaceallmissing=false)
```

Comment:
`indexcols` probably should not have type restriction at all or please give a standard restriction `Union{ColumnIndex, AbstractVector{Integer}, AbstractVector{Symbol}, Not, Regex}`.

Reply:
also `replacemissing` should be restricted to `Bool`, and `fill` does not need `::Any`